### PR TITLE
acmp: Specify SGID in PR query

### PR DESCRIPTION
**Pull request by mark-bloch**
As with commit: 03c383af0899a811e0f491699fc3a88c5708da26
There is a need to specify the SGID when issuing a PR query.

Today, we may have path (dest->path) with only the DGID part set.
Fix that by trying to use the source IP and find a matching interface.
Using acm_if_get_sgid to set the sgid accordingly.

Signed-off-by: Mark Bloch markb@mellanox.com

### CODE REVIEW

File: prov/acmp/src/acmp.c [hunk 1]
```diff
@@ -1774,6 +1845,7 @@ acmp_resolve_dest(struct acmp_ep *ep, struct acm_msg *msg, uint64_t id)
 		acm_log(2, "have address, resolving route\n");
 		acm_increment_counter(ACM_CNTR_ADDR_CACHE);
 		atomic_inc(&ep->counters[ACM_CNTR_ADDR_CACHE]);
+		acmp_acquire_sgid(saddr, dest);
```
Comment (shefty) [inline prov/acmp/src/acmp.c:1848]: This does not look right.  It seems that there should be a more generic solution that simply selects the source address, which could be a LID or GID.

Comment (shefty) [inline prov/acmp/src/acmp.c:1848]: This approach is assigning an SGID to the acmp_dest structure, which it then uses when sending a PR query.  However, the SGID association with the acmp_dest is temporary, and will be replaced the next time a call is made.

Either there needs to be a 1:1 correlation between an acmp_dest and the source address (SGID or SLID), or the source address needs to be passed to the PR query call.  _If_ there is a 1:1 correlation, then the source address should be assigned earlier in this flow, maybe when the acmp_dest is allocated and initialized.  If there is NOT a 1:1 relationship here, then the source addresss needs to be passed through and incorporated into the PR as appropriate.

Comment (mark-bloch) [inline prov/acmp/src/acmp.c:1848]: This is how it's done in the kernel today.

If you look at drivers/infiniband/core/addr.c:resolve_addr
We call rdma_copy_addr to store the dev_addr in struct rdma_dev_addr
which will later be used in cma_acquire_dev to match the device and be in included in the GRH (if a GRH is required).

Also the way the src ip is selected today, is to create and connect a UDP socket and take the source IP from the device the socket was bound to, so today already we have a connection between the ipoib interface and the PR query
I could've add the code in acm (acm_svr_select_src) but felt we should leave it up for the provider to select the SGID.

Comment (mark-bloch) [inline prov/acmp/src/acmp.c:1848]: > This approach is assigning an SGID to the acmp_dest structure, which it then uses when sending a PR query. However, the SGID association with the acmp_dest is temporary, and will be replaced the next time a call is made.
> 
> Either there needs to be a 1:1 correlation between an acmp_dest and the source address (SGID or SLID), or the source address needs to be passed to the PR query call. If there is a 1:1 correlation, then the source address should be assigned earlier in this flow, maybe when the acmp_dest is allocated and initialized. If there is NOT a 1:1 relationship here, then the source addresss needs to be passed through and incorporated into the PR as appropriate.

When we the receive the response, acmp_dest_sa_resp is called then is uses
the SGID to select the gid index and then marks : dest->state = ACMP_READY;
Which means next time the data we already have will be used.

Comment (shefty) [inline prov/acmp/src/acmp.c:1848]: I'm not objecting to converting a source IP/IPv6 address into an SGID.  However, the SGID that gets stored with the acmp_dest should not depend on which request gets processed by ibacm first.  That sort of assignment should be independent of any request.  I.e. how do you know that a second resolve operation uses the same IP address?  Or that the IP -> SGID mapping is the same?  How does this change handle something like: src=10.0.0.1 dst = 10.0.1.2 followed by src=10.0.0.2 dst=10.0.1.2?

It seems that either the acmp_dest mapping to an SGID is unique, in which case we can determine it earlier, such as when the acmp_dest is allocated, and we don't need the saddr as part of the process.  Or the acmp_dest mapping to an SGID is not unique, in which case, it isn't helpful to store the SGID with the acmp_dest.

Comment (shefty) [inline prov/acmp/src/acmp.c:1848]: Please also look at the other places where acmp_resolve_path_sa() is invoked and determine how the SGID is set in those locations.

Comment (mark-bloch) [inline prov/acmp/src/acmp.c]: > How does this change handle something like: src=10.0.0.1 dst = 10.0.1.2 followed by src=10.0.0.2 dst=10.0.1.2?

What happens today? aren't we storing the response from request A and reusing it for the
next request? (based on the dst ip and src ip)

acm_svr_resolve_dest is using the saddr (acm_get_ep_address) in order to get acmc_addr
so for different src address we will get a different acmp_addr which
will lead to a different acmp_ep which will lead to a different acmp_dest
( the call stack is: acm_svr_resolve_dest -> acmp_resolve -> acmp_resolve_dest)

I don't think we can assign the sgid when we allocate the acmp_dest, ibacm may be running
before ipoib is loaded into the kernel, which means there no ipoib interface from which we can take the dev_addr.

What do you think?

Comment (mark-bloch) [inline prov/acmp/src/acmp.c:1848]: For some reason it seems my comment was deleted, sorry about that, I'll write again.

> How does this change handle something like: src=10.0.0.1 dst = 10.0.1.2 followed by src=10.0.0.2 dst=10.0.1.2?

How is it done today?
in acm_svr_resolve_dest we use acm_svr_select_src to select the src addr
then we use  acm_get_ep_address (with the saddr) to find the acmc_addr
and then we jump into acmp

Which means different src address still work, for example:

```
markb@r-freeb14:~ $ ib_acme -V -d 1.1.2.13 -s 1.1.1.144
Service: localhost
Destination: 1.1.2.13
Source: 1.1.1.144
Path information
  dgid: fec0::2:14:500:0:19
  sgid: fec0::1:7cfe:9003:72:6f33
  dlid: 1
  slid: 2
  flow label: 0x0
  hop limit: 255
  tclass: 0
  reversible: 1
  pkey: 0xffff
  sl: 0
  mtu: 5
  rate: 12
  packet lifetime: 18

return status 0x0
markb@r-freeb14:~ $ ib_acme -V -d 1.1.2.13 -s 1.1.1.14
Service: localhost
Destination: 1.1.2.13
Source: 1.1.1.14
Path information
  dgid: fec0::2:14:500:0:19
  sgid: fec0::1:7cfe:9003:72:6f32
  dlid: 1
  slid: 3
  flow label: 0x0
  hop limit: 255
  tclass: 0
  reversible: 1
  pkey: 0xffff
  sl: 0
  mtu: 5
  rate: 12
  packet lifetime: 18

return status 0x0
```

I have a host file without a route file.
Which means I have an ip->gid mapping and that's it.

Now If I run ibacm before ipoib is loaded (and given an ip), I don't have any route rules
to be able to select the SGID based on the dst IP.

**Comment by shefty** [comment]
I don't see that my latest email response has made it here.  Re-posting relevant portion:

It looks like the traditional flow would have set the SGID as part of transitioning the acmp_dest state to ACMP_ADDR_RESOLVED.  See acmp_record_acm_addr() and acmp_record_path_addr().  The exception comes from acmp_parse_hosts_file(), which oddly enough has this code commented out:

```
//ibv_query_gid(ep->port->dev->verbs, ep->port->port_num,
//      0, &dest->path.sgid);
```

That looks like incorrect code that should just be removed.  Anyway, my suggestion would be to set the SGID in acmp_parse_hosts_file(), when the destination is allocated and the other fields (SLID, PKey, DGID) are set.  However, your argument is that this requires ipoib to be running, so that the routing tables are available...

Given that, we can defer mapping to an SGID, but using the source address from a user's request still _feels_ wrong.  I can't come up with a reasonable alternative to your approach though.

**Comment by mark-bloch** [comment]
> Given that, we can defer mapping to an SGID, but using the source address from a user's request still feels wrong. I can't come up with a reasonable alternative to your approach though.

How would you like to continue?
If you can come out with something better that would be great, I couldn't.
The _one_ thing that makes me feel _better_ is that we are doing the same process like the kernel.
i.e Using the dst ip to get the src interface and using dev_addr from that.

**Comment by shefty** [comment]
I was going to think about this some more, and if I can't come up with a better alternative soon, merge this.

**Comment by mark-bloch** [comment]
Sounds good to me, thank you for the help, if I can help somehow ping me.

**Pull request merged**